bootstore/src/schemes/v0/peer_networking.rs

```diff
@@ -0,0 +1,635 @@
+                    self.total_read = self.total_read - end;
+                    msg
+                }
+                Err(e) => {
+                    warn!(
+                        self.log,
+                        "Closing connection: failed to deserialize: {}", e
+                    );
+                    return self.close().await;
+                }
+            };
+            self.last_received_msg = Instant::now();
+            debug!(self.log, "Received {:?}", msg);
+            if let Msg::Fsm(msg) = msg {
+                if let Err(e) = self
+                    .main_tx
+                    .send(ConnToMainMsg {
+                        handle_unique_id: self.unique_id,
+                        msg: ConnToMainMsgInner::Received {
+                            from: self.peer_id.clone(),
+                            msg,
+                        },
+                    })
+                    .await
+                {
+                    warn!(
+                        self.log,
+                        "Failed to send received msg to main task: {:?}", e
+                    );
+                }
+            }
+        }
+    }
+
+    // Send ping messages and check for inactivity timeouts
+    async fn ping(&mut self) -> Result<(), ConnErr> {
+        if Instant::now() - self.last_received_msg > INACTIVITY_TIMEOUT {
+            warn!(self.log, "Closing connection: inactivity timeout",);
+            return self.close().await;
+        }
+        if self.write_queue.len() == MSG_WRITE_QUEUE_CAPACITY {
+            warn!(self.log, "Closing connection: write queue full",);
+            self.close().await
+        } else {
+            match write_framed(&Msg::Ping) {
+                Ok(msg) => {
+                    self.write_queue.push_back(msg);
+                    Ok(())
+                }
+                Err(e) => {
+                    warn!(
+                        self.log,
+                        "Closing connection: Failed to serialize msg: {}", e
+                    );
+                    self.close().await
+                }
+            }
+        }
+    }
+
+    // Close and drop the connection.
+    async fn close(&mut self) -> Result<(), ConnErr> {
+        if let Err(e) = self
+            .main_tx
+            .send(ConnToMainMsg {
+                handle_unique_id: self.unique_id,
+                msg: ConnToMainMsgInner::Disconnected {
+                    peer_id: self.peer_id.clone(),
+                },
+            })
+            .await
+        {
+            warn!(self.log, "Failed to send to main task: {:?}", e);
+        }
+        let _ = self.write_sock.shutdown().await;
+        Err(ConnErr::Retry)
+    }
+}
+/// Spawn a task that maintains a client connection to a peer
+pub async fn spawn_client(
+    unique_id: u64,
+    my_peer_id: Baseboard,
+    my_addr: SocketAddrV6,
+    addr: SocketAddrV6,
+    log: &Logger,
+    main_tx: mpsc::Sender<ConnToMainMsg>,
+) -> PeerConnHandle {
```

Comment:
Vague feelings:

I'm having a lot of trouble with the naming of `fn spawn_client(...) -> PeerConnHandle` and `fn spawn_server(...) -> AcceptedConnHandle`, which _both_ use an `EstablishedConn` internally in their task.

Maybe `ClientConnHandle` and `ServerConnHandle`?  Even with that name, it's hard to tell if it's a connection **to** a client or a connection in which **we** are taking the client role...

Similarly, do we want to split `EstablishedConn` into client / server flavors, since they handle different subsets of messages?  It seems like that would require different types for the `MainToConnMsg` / `ConnToMainMsg`, e.g. the client flavor would only send `Request` FSM messages and not `Response` (?).

Reply:
I think the names suck as discussed above. However, I don't think we should split the `EstablishedConn`. Once established both sides are equivalent. Either can send `Request` or `Response` messages.

Reply:
Fixed the names in c6735245f